utxo/src/utxo_impl/mod.rs

```diff
@@ -0,0 +1,537 @@
+                    // if utxo is spent and is not dirty, then it can be marked as fresh.
+                    !curr_entry.is_dirty || curr_entry.is_fresh
+                } else {
+                    // copy from the original entry
+                    curr_entry.is_fresh
+                }
+            }
+        };
+
+        // create a new entry
+        let new_entry = UtxoEntry::new(utxo, is_fresh, true);
+
+        // TODO: update the memory usage
+        // self.memory_usage should be added based on this new entry.
+
+        self.utxos.insert(key.clone(), new_entry);
+
+        Ok(())
+    }
+
+    /// Flags the utxo as "spent", given an outpoint.
+    /// Returns the Utxo if an update was performed.
+    pub fn spend_utxo(&mut self, outpoint: &OutPoint) -> Result<Utxo, Error> {
+        match self.get_utxo_entry(outpoint) {
+            None => Err(Error::NoUtxoFound),
+            Some(entry) => {
+                let key = outpoint;
+
+                // TODO: update the memory usage
+                // self.memory_usage must be deducted from this entry's size
+
+                // check whether this entry is fresh
+                if entry.is_fresh {
+                    // This is only available in this view. Remove immediately.
+                    self.utxos.remove(key);
+                } else {
+                    // mark this as 'spent'
+                    let new_entry = UtxoEntry {
+                        status: UtxoStatus::Spent,
+                        is_dirty: true,
+                        is_fresh: false,
+                    };
+                    self.utxos.insert(key.clone(), new_entry);
+                }
+
+                entry.take_utxo().ok_or(Error::UtxoAlreadySpent)
+            }
+        }
+    }
+
+    /// Checks whether utxo exists in the cache
+    pub fn has_utxo_in_cache(&self, outpoint: &OutPoint) -> bool {
+        let key = outpoint;
+        self.utxos.contains_key(key)
+    }
+
+    /// Returns a mutable reference of the utxo, given the outpoint.
+    pub fn get_mut_utxo(&mut self, outpoint: &OutPoint) -> Option<&mut Utxo> {
+        self.get_utxo_entry(outpoint).and_then(|status| {
+            match status.status {
+                UtxoStatus::Spent => None,
+                UtxoStatus::Entry(utxo) => {
+                    let key = outpoint;
+                    self.utxos.insert(
+                        key.clone(),
+                        UtxoEntry::new(utxo, status.is_fresh, status.is_dirty),
+                    );
+                    //TODO: update the memory storage here
+                    self.utxos.get_mut(key).and_then(|entry| entry.utxo_mut())
+                }
+            }
+        })
+    }
+
+    /// removes the utxo in the cache with the outpoint
+    pub(crate) fn uncache(&mut self, outpoint: &OutPoint) -> Option<UtxoEntry> {
+        let key = outpoint;
+        if let Some(entry) = self.utxos.get(key) {
```

Comment:
This may be possible to simplified by combining `filter()` and `map()`, not sure if it's worth it though.

Reply:
Yeah, many things are not optimum in the PR. It's not a priority though. We'll fix the little things we can.

Reply:
The source of these things is that Carla copied bitcoin, btw.